common/browser_process.go

```diff
@@ -181,49 +181,69 @@ func execute(
 
 // parseDevToolsURL grabs the WebSocket address from Chrome's output and returns
 // it. If the process ends abruptly, it will return the first error from stderr.
-func parseDevToolsURL(ctx context.Context, cmd command) (string, error) { //nolint: cyclop
-	var (
-		lines = make(chan string)
-		errCh = make(chan error)
-	)
+func parseDevToolsURL(ctx context.Context, cmd command) (_ string, err error) {
+	parser := &devToolsURLParser{
+		sc: bufio.NewScanner(cmd.stderr),
+	}
+	done := make(chan struct{})
 	go func() {
-		scanner := bufio.NewScanner(cmd.stderr)
-		for scanner.Scan() {
-			lines <- scanner.Text()
-		}
-		if err := scanner.Err(); err != nil {
-			errCh <- err
+		for parser.scan() {
 		}
+		close(done)
 	}()
+	for err == nil {
```

Comment:
Why is this a for loop?

Reply:
Hmm 🤔 Sometimes, the browser process can only return an empty string. And the scanner will return `false` (_no more lines to read_). So, the `err` from the `case <-done: err = parser.err()` will be `nil`. The loop waits until the command (`<-cmd.done`) finishes and returns with an error (or the context (<- `ctx.done`).

We're waiting for the command to exit so that we can run this:

https://github.com/grafana/xk6-browser/blob/6c3449f1d5ac697b21a66d3c7215e862ea51e919/common/browser_process.go#L162-L177

The `done` channel there is closed when we successfully delete the user data directory.